Add optional content checking to ResourceWatcher

tvernum opened the pull request:
In some cloud environments, there may be frequent synchronization of
configuration files from the orchestration layer to the ES container.

This can trigger frequent, unnecessary reloading of files.

Previously, code that used the ResourceWatcherService / FileWatcher
would need to detect "no-op" file changes itself. With the addition of
this content checking support, it can be handled efficiently by the
Resource Watcher Service.

tvernum committed:
Add optional content checking to ResourceWatcher

In some cloud environments, there may be frequent synchronization of
configuration files from the orchestration layer to the ES container.

This can trigger frequent, unnecessary reloading of files.

Previously, code that used the ResourceWatcherService / FileWatcher
would need to detect "no-op" file changes itself. With the addition of
this content checking support, it can be handled efficiently by the
Resource Watcher Service.

elasticmachine commented:
Pinging @elastic/es-core-infra (Team:Core/Infra)

rjernst reviewed (approved):
LGTM

tvernum committed:
Merge branch 'master' into file-watcher-hash

ywangd reviewed (approved):
LGTM

This might be a good tradeoff in the situation when there are many more false postives (watcher gets notified when file does not actually change). But it adds an additional file read for true postives. Also the downstream consumer still needs to check for false postives though much less likely to happen. Overall since false postives are much more likely to happen and true postives are relatively rare, it's an improvement.

tvernum committed:
Address feedback

- Calculate hash on first file read
- Calculate hash everytime file size changes
- Additional tests
